docs/GraphRepresentation.md

```diff
@@ -0,0 +1,321 @@
+
+The data used in generating recommendations can be summarized
+by such a graph, where
+an add-on's candidate list corresponds to its set of neighbours in the graph,
+and relevance scores are given by the edge weights.
+In this setting:
+
+- A relational graph is induced directly by the coinstallation data,
+    considering add-ons related if they are coinstalled,
+    and using coinstallation counts as weights.
+- Treatments can be thought of as transformations
+    which modify add-on relational graphs by
+    adding or deleting edges, or adjusting the edge weights.
+- The final selection of recommendations can itself be viewed as a treatment,
+    which outputs a graph containing, for each add-on A,
+    an edge from A to each of A's N recommendations.
+
+The recommendation algorithm can be rephrased
+in terms of this representation as follows:
+
+1. Load the input data and create the __coinstallation graph__.
+2. Apply each of the specified treatments in turn to the coinstallation graph,
+    resulting in the __treated graph__.
+3. Apply the recommendation selection treatment to the treated graph
+    to produce the __recommendation graph__.
+
+Note that, since the graph is directed, edges do not necessarily run both ways.
+In other words, we allow for a treatment to produce a graph in which
+A is related to B but B is not related to A,
+for example if it removes edges whose weight falls below a threshold.
+However, since the relation of being coinstalled is symmetric,
+the initial coinstallation graph is undirected.
+
+A relational graph has an associated __adjacency matrix__ $C$,
+in which rows and columns are indexed by add-ons (graph vertices),
+and entry $C_{ij}$ contains the weight for edge $(i,j)$
+if these two add-ons are connected in the graph, or 0 otherwise.
+The initial coinstallation dataset is in fact stored
+as a sparse representation of its adjacency matrix.
+Since this graph is undirected, the adjacency matrix is symmetric
+($C_{ij} = C_{ji}$ for any pair $i,j$).
+Also, the row sum (or column sum, by symmetry) for add-on $i$ gives
+its overall number of installs across all profiles considered in the dataset.
+
+
+## Graph properties
+
+For reference, we now list some of the main properties of
+each of the milestone relational graphs listed above.
+
+
+### Coinstallation graph
+
+The graph induced by the raw coinstallation counts has the following properties:
+
+- It is undirected, ie. each edge runs both ways.
+- Its vertex degree may take values from 1 to $|V|$,
+    the overall number of add-ons in the graph.
+    In other words, every vertex has at least one incident edge.
+- Edge weights correspond to the raw coinstallation count.
+- It may or may not be connected.
+    However, there are no unreachable (singleton) vertices.
+
+
+### Treated graph
+
+The result of applying the treatments to the coinstallation graph
+has these properties:
+
+- It is directed, as treatments generally do not operate on edges symmetrically.
```

Comment:
I softened this a bit. Some treatments could leave it undirected, but even pruning could operate asymmetrically by removing an edge in one direction but not another (eg. dropping low-scoring recommendation candidates).

The idea is that, in general, the treated graph is directed. In some cases it could be undirected, but that's still a special case of directed, so any properties we discuss would still apply.